kernel/src/scan/mod.rs

```diff
@@ -441,6 +450,133 @@ impl Scan {
+                    StructField::not_null("path", DataType::STRING),
+                    StructField::not_null("partitionValues", partition_values),
+                    StructField::not_null("size", DataType::LONG),
+                    StructField::nullable("modificationTime", DataType::LONG),
+                    StructField::nullable("stats", DataType::STRING),
+                    StructField::nullable("deletionVector", DeletionVectorDescriptor::to_schema()),
+                ]),
+            )])
+        });
+
+        // TODO(#966): validate that the current predicate is compatible with the hint predicate.
+
+        if hint_version > self.snapshot.version() {
+            return Err(Error::Generic(format!(
+                "hint_version {} is greater than current version {}",
+                hint_version,
+                self.snapshot.version()
+            )));
+        }
+
+        // in order to be processed by our log replay, we must re-shape the existing scan metadata
+        // back into shape as we read it from the log. Since it is already reconciled data,
+        // we treat it as if it originated from a checkpoint.
+        let transform = engine.evaluation_handler().new_expression_evaluator(
+            scan_row_schema(),
+            get_scan_metadata_transform_expr(),
+            RESTORED_ADD_SCHEMA.clone(),
+        );
+        let apply_transform =
+            move |data: Box<dyn EngineData>| Ok((transform.evaluate(data.as_ref())?, false));
+
+        // If the snapshot version corresponds to the hint version, we process the existing data
+        // to apply file skipping and provide the required transformations.
+        if hint_version == self.snapshot.version() {
+            let scan = hint_data.into_iter().map(apply_transform);
+            return Ok(Box::new(self.scan_metadata_inner(engine, scan)?));
+        }
+
+        let log_segment = self.snapshot.log_segment();
+
+        // If the current log segment contains a checkpoint newer than the hint version
+        // we disregard the existing data hint, and perform a full scan. The current log segment
+        // only has deltas after the checkpoint, so we cannot update from prior versions.
+        if matches!(log_segment.checkpoint_version, Some(v) if v > hint_version) {
+            return Ok(Box::new(self.scan_metadata(engine)?));
+        }
+
+        // create a new log segment containing only the commits added after the version hint.
+        let mut ascending_commit_files = log_segment.ascending_commit_files.clone();
+        ascending_commit_files.retain(|f| f.version > hint_version);
+        let listed_log_files = ListedLogFiles {
+            ascending_commit_files,
+            ascending_compaction_files: vec![],
+            checkpoint_parts: vec![],
+        };
+        let new_log_segment = LogSegment::try_new(
+            listed_log_files,
+            log_segment.log_root.clone(),
+            Some(log_segment.end_version),
+        )?;
```

Comment:
flagging that we are starting to have a lot of different use cases where we need to modify LogSegments or do special listing, etc. (here and Snapshot::try_new_from, etc.) - perhaps should spend some time exploring a nicer api?